Proposal: Mapping REST to Graphql (#612)
kyma-incubator/compass: docs/investigations/rest-adapters/rest-adapter-for-graphql.md

# Provide REST Adapter for Compass Graphql API
Compass GraphQL API is not compatible with Application Connector REST API.

> To avoid rewriting integration layer for  Application Connector clients, we can provide adapter that translates

Reviewer: ```suggestion
To avoid rewriting integration layer for  Application Connector clients, we can provide Adapter that translates
```

the status of TokenRequest will be populated differently, depending on whether the Application CR comes from Agent or was created manually.

AFAIK, creating an Application was a manual step usually performed in UI. Because of that, we should consider
not introducing previously mentioned controller, and just instruct Admins to create an Application using Compass UI. 

2. Get a client certificate

When creating a TokenRequest, the Controller updates the status and provides token and configuration URL fields.
There is an open question, if we should support getting token by creating TokenRequest.
If so, Agent should be responsible with setting status of TokenRequest.

After calling the `/v1/applications/signingRequests/info` endpoint, the following data is returned:
```json

{
    "csrUrl": "",
    "api":{
        "metadataUrl":      "",
        "eventsUrl":        "",
        "infoUrl":          "",
        "certificatesUrl":  "",
    },
    "certificate":{
        "subject":"OU=Test,O=TestOrg,L=Waldorf,ST=Waldorf,C=DE,CN={APP_NAME}",
        "extensions": "",
        "key-algorithm": "rsa2048",
    }
}
```

> `csrUrl` should point to the adapter's `/v1/applications/certificates`.

Reviewer: ```suggestion
`csrUrl` should point to the Adapter's `/v1/applications/certificates`.
```

described [here](https://kyma-project.io/docs/1.8/components/application-connector/#tutorials-tutorials)
still works.

1. Create a new Application

In the old approach, Application CR was created manually in the Runtime. In the new approach, Agents read information from Compass 
Director and creates appropriate Application CRs.
There is an open question, what should happen for manually created Application CRs. Right now, Agent ignores them completely.
If there is a requirement, that manually created Application should be automatically registered in the Compass, 
a new Controller has to be implemented.
If there is a requirement, that manually created Application CR should be handled by old services, this complicates implementation, 
because status of TokenRequest will be populated differently, depending if Application CR comes from Agent or was created manually.

AFAIK, creating application was a manual step usually performed in UI. Because of that, we should consider
not introducing previously mentioned controller, and just instruct Admins to create application using Compass UI. 

2. Get the client certificate

> When creating Token Request, a  Controller updates the status and provide token and configuration URL fields.

Reviewer 2: I'm not sure if this component is still valid with the Compass scenarios and approach. The required changes actually depend on whether we support the creation of the applications on the runtimes.

If we go with the limitation that all applications must be registered via the compass then this controller needs to call the Compass API to obtain the token and it may be worth to just update it to support just the new GraphQL API. In this case we need to consider how this component will communicate with the Compass instance - maybe it should be replaced by the Runtime Agent (?) which already has a trusted relation established.

```json

{
    "csrUrl": "",
    "api":{
        "metadataUrl":      "",
        "eventsUrl":        "",
        "infoUrl":          "",
        "certificatesUrl":  "",
    },
    "certificate":{
        "subject":"OU=Test,O=TestOrg,L=Waldorf,ST=Waldorf,C=DE,CN={APP_NAME}",
        "extensions": "",
        "key-algorithm": "rsa2048",
    }
}
```

`csrUrl` should point to the adapter's `/v1/applications/certificates`.
`certificate` field is populated by adapter by calling Connector's GraphQL query `configuration`. 

3. Register a service

```
curl -k --cert ./app1_client.crt   --key ./generated.key -X POST https://gateway.34.77.12.120.xip.io/app1/v1/metadata/services --data-binary "@serviceRegistrationManyAPIs.json"
```
```json
{
  "provider":"SAP",
  "name":"api1",
  "description": "api1 desc",
  "shortDescription": "api1 short desc",
  "identifier": "api1Identifier",
  "labels":{
  	"label1":"value1",
  	"label2": "value2"
  },
  "api":{
  	"targetUrl":"http://some-app1.pl",
  	"requestParameters": {
  	  "headers": {
  	      "header1": "value1"
  	  }
  	}
  }

}
```
Compass equivalent: `addApiDefinition` or `addEventDefinition`

Challenges:

> - in old API, there is a Service which seems to be equivalent to API Package, which is missing in Compass

Reviewer: ```suggestion
- in the former API, there is a service which seems to be equivalent to API Package that is missing in Compass
```

> - `identifier` has no mapping in Compass API

Reviewer: ```suggestion
- `identifier` has no mapping in the Compass API
```

> - you cannot label service class via Compass API

Reviewer: ```suggestion
- you cannot label ServiceClass using Compass API
```

- specificationURL, specificationCredentials, specificationRequestParameters can be mapped to FetchRequest, but it is not yet implemented

> - from Documentation:

Reviewer: ```suggestion
- quoting the documentation:
```

> If the api.spec or api.specificationUrl parameters are not specified and the api.type parameter is set to OData, the Application Registry will try to fetch the specification from the target URL with the $metadata path.

> Special handling for OData can be implemented in the Adapter (assuming that FetchRequest is implemented)

Reviewer: ```suggestion
Special handling for OData can be implemented in the Adapter (assuming that FetchRequest is implemented).
```

> TODO: CSRF tokens are not implemented in Compass

Reviewer: ```suggestion
TODO: 
- CSRF tokens are not implemented in Compass
```

> - documentation: in Compass, documentation can be added only on the Application level

Reviewer: ```suggestion
- Documentation: in Compass, documentation can be added only on the Application level
```

> # Diagram

Reviewer 2: The adapters are not a part of the Kyma Runtime. They are components that are run inside the Management Plane. I would put them in front of the GraphQL API components.

If we go with the TokenRequest CR then it is a part of the Kyma Runtime.

![](adpters.svg)

> Kyma Integration Layer implemented as a part of the External Solution reads status of the TokenRequest, which returns URL pointing to Adapters.

Reviewer: ```suggestion
Kyma Integration Layer, implemented as a part of the external solution, reads the status of the TokenRequest that returns a URL pointing to Adapters.
```

> Adapters proxies calls to Compass components (REST --> GraphQL) and performing type conversions. 

Reviewer: ```suggestion
Adapters proxy calls to Compass components (REST --> GraphQL) and perform type conversions. 
```

> Still open question is how to propagate credentials, for example Adapter endpoint is reached using client certificate, can we propagate

Reviewer: ```suggestion
Still, the open question is how to propagate credentials. For example, the Adapter endpoint is reached using a client certificate. Can we propagate
```
